pkg/controller/resourceclaim/controller.go

```diff
@@ -854,12 +857,52 @@ func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) err
 		// those, the resource claim controller will trigger deletion when the
 		// pod is done. However, it doesn't hurt to also trigger deallocation
 		// for such claims and not checking for them keeps this code simpler.
-		if len(valid) == 0 &&
-			claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer {
-			claim.Status.DeallocationRequested = true
+		if len(valid) == 0 {
+			if builtinControllerFinalizer >= 0 {
+				if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer ||
+					claim.DeletionTimestamp != nil {
+					// Allocated by scheduler with structured parameters. We can "deallocate"
+					// by clearing the allocation.
+					claim.Status.Allocation = nil
+				}
+			} else if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer {
+				// DRA driver controller in the control plane
+				// needs to do the deallocation.
+				claim.Status.DeallocationRequested = true
+			}
+			// In all other cases, we keep the claim allocated, in particular for immediate allocation
+			// with a control plane controller.
+		}
+
+		claim, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+		if err != nil {
+			return err
 		}
 
-		_, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{})
+		// Now also remove the finalizer. Note that the index may have changed.
+		builtinControllerFinalizer := slices.Index(claim.Finalizers, resourcev1alpha2.Finalizer)
+		if builtinControllerFinalizer >= 0 {
+			claim.Finalizers = slices.Delete(claim.Finalizers, builtinControllerFinalizer, builtinControllerFinalizer+1)
+			if _, err := ec.kubeClient.ResourceV1alpha2().ResourceClaims(claim.Namespace).Update(ctx, claim, metav1.UpdateOptions{}); err != nil {
+				return err
+			}
+		}
```

Comment:
These 2 updates seems a bit excessive. Why not just do a single update once here? What's the benefit of doing status update a few lines above. Even though you're invoking update status only, in both cases a full resource claim is being send as the body. So why not just do it once?